internal/pipeline/executor.go

```diff
@@ -207,78 +179,84 @@ func (s hydrationState) String() string {
 	return []string{"Dry", "Hydrating", "Wet"}[s]
 }
 
-//
-// hydrate hydrates a pkg and returns wet resources.
-func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error) {
-	state, found := hctx.hydrationStates[p.Path()]
+// hydrate hydrates given pkg and returns wet resources.
+func hydrate(pn *pkgNode, hctx *hydrationContext) (resources []*yaml.RNode, err error) {
+	currPkg, found := hctx.pkgs[pn.Path()]
 	if found {
-		switch state {
+		switch currPkg.state {
 		case Hydrating:
-			// we have a cycle
-			// TODO(droot): improve the error to contain useful information
-			err = fmt.Errorf("found cycle in dependencies for pkg %s", p.Path())
+			// we detected a cycle
+			err = fmt.Errorf("found cycle in dependencies for pkg %s", currPkg.Path())
+			return resources, err
 		case Wet:
-			// already hydrated, we sort of assume pkgs in hctx will be populated but we
-			// can be defensive here
-			resources = hctx.pkgs[p.Path()].resources
-		default:
-			err = fmt.Errorf("invalid pkg state %v detected", state)
+			resources = currPkg.resources
+			return resources, err
+		case Dry:
+			// Is this a dead path ?
 		}
-		return resources, err
+	} else {
+		// add it to the discovered package list
+		hctx.pkgs[pn.Path()] = pn
+		currPkg = pn
 	}
 	// mark the pkg in hydrating
-	hctx.hydrationStates[p.Path()] = Hydrating
-	// add to the graph
-	hctx.pkgs[p.Path()] = p
-
-	// read the pkg, the pipeline, dependencies etc.
-	err = p.Read()
-	if err != nil {
-		err = fmt.Errorf("failed to read the pkg %w", err)
-		return resources, err
-	}
+	currPkg.state = Hydrating
 
 	// input resources for current pkg's hydration
 	var input []*yaml.RNode
-	// here we explore the p.sources to determine sub packages etc.
-	sources, err := p.resolveSources()
+	// determine sub packages to be hydrated
+	pkgPaths, err := currPkg.subPackages()
 	if err != nil {
 		return resources, err
 	}
-	for _, s := range sources {
-		// TODO(droot): sync with pipeline's defaults once Donny's PR is merged
-		if s == p.Path() {
+	klog.Infof("hydrating pkg %s with sub packages %v", currPkg.Path(), pkgPaths)
+	for _, s := range pkgPaths {
+		if s == currPkg.Path() {
 			var currPkgResources []*yaml.RNode
 			// current pkg will be hydrated as we drop of this loop
-			currPkgResources, err = p.localResources(false)
+			currPkgResources, err = currPkg.localResources(false)
 			if err != nil {
 				return resources, err
 			}
+			// Resources are read from local filesystem at a package level, so the
+			// path annotation in each resource points to path relative to that package.
+			// But the resources are written to the file system at the root package level, so
+			// the path annotation in each resources needs to be adjusted to be relative to the rootPkg.
+			relPath, err := filepath.Rel(hctx.root.Path(), s)
```

Comment:
This is the kind of low-level operation that needs to be done once and reused instead of re-implemented multiple times.

Also how do you handle new resources generated in the package after executing the pipeline? Please compare to to how 'kpt fn run' works today.

Reply:
I tweaked this to handle generated resources and also added a basic testcase for generator in a nested package. Also changed the order, it is not being done as a post-hydration step at the pkg level. I can potentially move it to a helper method now.